### DISCUSSION TITLE
Distance coefficient becomes zero when I set a crs

### DISCUSSION THREAD

**Discussion by InaKrapp** [inlabru-org/inlabru#257] [Q&A] [answered]
Hello,
I have a dataset of points in a city for which I aim to create a model which uses the distance to the city center as explanatory variable.
But when I run a model, it gives me a coefficient of zero with a standard deviation of zero for the distance. 

I tried to isolate the issue. In a small testdataset where I did not set a crs, I obtained a nonzero coefficient and a positive standard deviation, but it became zero when I used the same values to create a dataset where I set a crs.
So I think it may be related to the crs, but I don't understand how or why.
Here is the example data where I do not set a crs:

```
#Create example data: 
m1 = cbind(c(0, 0, 1, 0), c(0, 1, 1, 0))
m2 = cbind(c(0, 1, 1, 0), c(0, 0, 1, 0))
pol = st_sfc(st_polygon(list(m1)), st_polygon(list(m2)))
set.seed(1985)
d = data.frame(matrix(runif(15), ncol = 3))
testdata_working = st_as_sf(x = d, coords = 1:2)
# Calculate distance
testdata_working$dist = st_distance(testdata_working, st_point(c(0.5,0.5)))
# Plot example data
ggplot() + geom_sf(aes(geometry = geometry, color = dist), data = testdata_working)

# Create and plot mesh
max.edge = diff(range(st_coordinates(testdata_working)[,1]))/(3*5)
mesh2 = inla.mesh.2d(loc=st_coordinates(testdata_working), max.edge = c(1,2)*max.edge)
ggplot() +  gg(mesh2) +  geom_sf(data=testdata_working,col='purple',size=1.7,alpha=0.5)

# This toy example is also working fine if I don't run this line.
mesh2$crs <- st_crs(testdata_working)

# Fit a simple INLA model
matcov = inla.spde2.pcmatern(mesh2, prior.sigma = c(0.1, 0.01), prior.range = c(0.1, 0.01)) 
mdl= geometry~ dist
testfit_working = lgcp(mdl, testdata_working, domain = list(geometry = mesh2))
summary(testfit_working)
```
Running this gives me the following output: 
<img width="410" alt="inlabru_working_example" src="https://github.com/user-attachments/assets/83f4281e-9a82-4fce-a8c9-5a744f191763">

But if I create a second example dataset with exactly the same values where I add a crs, the coefficient and the standard deviation of the distance variable become zero.
Here is the code for the second example:
```
# Create example data
m1 = cbind(c(0, 0, 1, 0), c(0, 1, 1, 0))
m2 = cbind(c(0, 1, 1, 0), c(0, 0, 1, 0))
pol = st_sfc(st_polygon(list(m1)), st_polygon(list(m2)))
set.seed(1985)
d = data.frame(matrix(runif(15), ncol = 3))
testdata_failing = st_as_sf(x = d, coords = 1:2, crs = "EPSG:4326" )
point = st_sf(a = "point", geometry = st_sfc(st_point(c(0.5,0.5))))
st_crs(point) = "EPSG:4326"
testdata_failing$dist = as.numeric(st_distance(testdata_failing, point ))
ggplot() + geom_sf(aes(geometry = geometry, color = dist), data = testdata_failing)

# Create mesh
max.edge = diff(range(st_coordinates(testdata_failing)[,1]))/(3*5)
mesh2 = inla.mesh.2d(loc=st_coordinates(testdata_failing), max.edge = c(1,2)*max.edge)
ggplot() +  gg(mesh2) + geom_sf(data=testdata_failing,col='purple',size=1.7,alpha=0.5)
mesh2$crs <- st_crs(testdata_failing)

# Fit a simple INLA model
matcov = inla.spde2.pcmatern(mesh2, prior.sigma = c(0.1, 0.01), prior.range = c(0.1, 0.01))  
mdl= geometry~ dist
testfit_failing = lgcp(mdl, testdata_failing,  domain = list(geometry = mesh2))
summary(testfit_failing)
```
Here is a screenshot of the output I obtained for this code:
<img width="418" alt="inlabru_failing_example" src="https://github.com/user-attachments/assets/61bbe8d1-077c-4c0c-baed-a47cf054d998">

Any help would be very apprechiated.

**Comment by finnlindgren** [comment, marked as answer]
EPSG:4326 is raw longlat, for which sf handles things differently than for projected coordinates, so that's the likely root cause of the problem; the units as seen by fmesher is in the local projection units, whereas sf with longlat calls s2 to convert distances to metres.
In toy problems like your example code, the crs isn't a real thing, but whenever you're using a crs, it has to make sense geographically. (Also see below regarding modelling in longlat; in short, don't!)

Also, you should not set the mesh crs manually after creating it. Instead, supply the desired mesh crs to the meshing call (`fm_mesh_2d_inla()`, which replaced `inla.mesh.2d()`), and avoid `st_coordinates()` where not absolutely necessary; If you give input points in sf with appropriate crs, the internal code will convert them to the target crs when necessary; it cannot do this if you've removed the points from their crs context.

So check the actual distance values you're using, and the actual coordinate values of the mesh. Also, It's almost never appropriate to build spde models in longlat coordinates, as that induces artificial axis-aligned anisotropy. Use a local projection for your study region instead. Usually that has to use km as units as well, to avoid numerical issues with huge metre values.
Set the units like this, assuming mycrs is first set to be an appropriate projection crs:
```
fm_length_unit(mycrs) <- "km"
```

**Reply by finnlindgren** [reply]
Pretty sure it's partly a units issue, and converting to as sensible crs solves parts of it. It's not just the mesh that should be in a sensible crs. The units of the distance values need to be sensible as well. Your last example still uses longlat for the data, so the distances will be in metres due to sf using s2 to calculate distances.

But you're also not supplying a proper model definition; for lgcp, you must construct the components to evaluate as functions of the point process domain, here `geometry`. Not sure what actually goes in the likelihood in your case, so the estimates are uninterpretable even without the other issues.

**Comment by InaKrapp** [comment]
Thank you very much for your quick answer. Unfortunately, I have not been able to resolve my problem fully: 

1. Yes, the toy example used a raw longlat projection, that was my error there. But the problem also occurs with my original data, which uses the 'EPSG:25833' projection instead. It does not occur with the toy data if I assign the 'EPSG:25833' projection to the toy data, though.
2. I updated the code to create the mesh like you suggested. I created a new mesh using the code:
`mesh = fm_mesh_2d_inla(loc=data, crs = st_crs(data))`
for my actual dataset named 'data'. But even though using 'st_crs' tells me the dataset has the expected crs, and `mesh$crs` also returns the same crs, I get an error when I try to run the model again: 
'Error in fm_transform_sf(x, crs = crs, ..., passthrough = passthrough) : 
  'crs' is an invalid coordinate reference object.'. 
Do you have any idea what may be causing this?

Best Wishes and sorry for the trouble.

**Reply by finnlindgren** [reply]
What's `fm_crs(data)` and `fm_crs(mesh)`? (note: `fm_crs`, not `st_crs`)

(normally, `fm_crs` would simply call `st_crs` for sf objects, but `fm_crs` works for meshes as well)

There may be several different issues here; if anything is in raw longlat, that would cause issues whenever using sf in combination with fmesher, which would then explain why the toy example works when it's not longlat.

I wouldn't really want to debug issues with known problematic uses of sf, so please make the test case use something other than longlat. Either a non-longlat crs, or an NA crs.

The `traceback()` output after the error message would be helpful. Without it, there's no way to tell what part of the code triggers the error.

(Sorry about the multi-answer; have aggregated into a single entry now)

**Comment by InaKrapp** [comment]
Hello, and sorry for the late reply. 
Yes, fm_crs() gave different results for the data and the mesh. I ran the code to adjust the distance of the data's crs, 'fm_length_unit(mycrs) <- "km"' after I had already defined the mesh, so it changed the crs of the data, but not the crs of the mesh. Therefore, I got the error because the crs was not identical anymore. Thank you for the remark to use 'fm_crs' to check, that helped me solve this issue.

The issue with the coefficient becoming zero in my dataset with the projected crs turned out to be related to the aspect you also mentioned earlier: I used a local projection with metres as units, which appears to have caused numerical issues. Since this is a characteristic of the dataset, I haven't encountered the same issue in the test dataset.

Thank you very much for your help!
